Split extra vars for Ansible playbooks

fdupont-redhat opened the pull request:
With https://github.com/ManageIQ/manageiq/pull/19698, we switched to using AwesomeSpawn to build the command lines, including `ansible-playbook`. So far, we used a single `--extra-vars` option with a JSON string, but the command line generated by AwesomeSpawn is broken, probably by too much escaping.

Given that `extra_vars` is a hash where each value is a string, this pull requests iterates over the hash entries to add them as independent `--extra-vars` option. This way, there no conflict between the generated JSON string and the escaping of AwesomeSpawn.

RHBZ: https://bugzilla.redhat.com/show_bug.cgi?id=1806485

fdupont-redhat commented:
@miq-bot add-label bug, v2v, ivanchuk/yes, blocker
@miq-bot add-reviewer @Fryguy

Fryguy commented:
> So far, we used a single --extra-vars option with a JSON string, but the command line generated by AwesomeSpawn is broken, probably by too much escaping.

While I don't mind this PR, and it's probably cleaner to do it this way, can you give an example of a broken value?  I tried all kinds of combinations with this and never had a problem.

fdupont-redhat commented:
When the playbook fails to run, here is the command line dumped in the evm.log:

```
ansible-playbook /usr/share/v2v-conversion-host-ansible/playbooks/conversion_host_enable.yml --become --inventory 10.8.59.14, --extra-vars=ansible_ssh_common_args\=\'-o\ StrictHostKeyChecking\=no\' --user cloud-user --private-key /tmp/ansible_key20200224-8377-6xxgoy --extra-vars \'\{\"v2v_host_type\":\"openstack\",\"v2v_transport_method\":\"vddk\",\"v2v_vddk_package_url\":\"http://file.rdu.redhat.com/\~ytale/VMware-vix-disklib-6.5.2-6195444.x86_64.tar.gz\"\}\'
```

The error message from the playbook is: `'v2v_transport_method' is undefined`.

djberg96 commented on app/models/conversion_host.rb:
```diff
@@ -454,7 +454,7 @@ def ansible_playbook(playbook, extra_vars = {}, miq_task_id = nil, auth_type = '
     end
 
-    params << {:extra_vars => "'#{extra_vars.to_json}'"}
+    extra_vars.each { |k, v| params << {:extra_vars= => "#{k}='#{v}'"} }
```
It's ok to have multiple `:extra_vars`?

fdupont-redhat replied:
@djberg96 yes, it's perfectly fine: https://docs.ansible.com/ansible/2.9/user_guide/playbooks_variables.html#passing-variables-on-the-command-line

Fryguy commented:
I don't understand why that value would fail...it looks perfect.

Fryguy commented:
Gah, I'm sorry...I totally misread the OP.  When I saw `we switched to using AwesomeSpawn`, I read it as `we switched to using AnsibleRunner`, and was so confused how the built-in runner implementation could have been wrong.

fdupont-redhat commented:
> Gah, I'm sorry...I totally misread the OP. When I saw `we switched to using AwesomeSpawn`, I read it as `we switched to using AnsibleRunner`, and was so confused how the built-in runner implementation could have been wrong.

AnsibleRunner is the next step :)

Fryguy commented on app/models/conversion_host.rb:
```diff
@@ -454,7 +454,7 @@ def ansible_playbook(playbook, extra_vars = {}, miq_task_id = nil, auth_type = '
     end
 
-    params << {:extra_vars => "'#{extra_vars.to_json}'"}
+    extra_vars.each { |k, v| params << {:extra_vars= => "#{k}='#{v}'"} }
```
I'm going to merge this for now since it's working for you, but I still think it's wrong.  For example, it manually embeds single quotes around the value without taking care of whether single quotes are actually in the strings being embedded.  However it was like that previously as well, so it's no different with respect to this PR.

fdupont-redhat replied:
Thanks. When we move to AnsibleRunner, I guess this won't be a topic anymore.

simaishi commented:
Ivanchuk backport details:

```
$ git log -1
commit 783357be593cbb5b156562fb47224c9a39b1f469
Author: Jason Frey <jfrey@redhat.com>
Date:   Mon Feb 24 11:39:35 2020 -0500

    Merge pull request #19864 from fdupont-redhat/v2v_conversion_host_fix_extra_vars

    Split extra vars for Ansible playbooks

    (cherry picked from commit d61bf49cbdc7fd6fc3fac2fc53466bffa165a6a3)

    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1806643
```

simaishi commented:
Reverted

```
commit bc5a20cc0efe751095482564bee1137a7a04753e
Author: Satoe Imaishi <simaishi@redhat.com>
Date:   Mon Mar 9 13:57:35 2020 -0400

    Revert "Merge pull request #19864 from fdupont-redhat/v2v_conversion_host_fix_extra_vars"

    This reverts commit 783357be593cbb5b156562fb47224c9a39b1f469.

    https://bugzilla.redhat.com/show_bug.cgi?id=1806643
```

simaishi commented:
This was reverted in #20168 for master/jansa.
